### PR TITLE
Add Reader.LookupOffset & Decode

### PR DESCRIPTION
**Pull request by jgraettinger**
Expose database offsets, and decoding records at those offsets, as a
first-class concept via LookupOffset & Decode. Allow decoded records to
indirect nested structures by introducing special handling for the
uintptr type: fields with this type are decoded by storing the database
offset of the indirected record.

Reflection is expensive, many database records ("continent", "country", etc)
are embarrassingly cacheable, and database offsets make for excellent caching
keys. These methods allow clients to trivially identify and re-use structures
which have previously been extracted.

---
This change is [Reviewable](https://reviewable.io/reviews/oschwald/maxminddb-golang/23)

### CODE REVIEW

File: decoder.go [hunk 1]
```diff
@@ -247,8 +253,15 @@ func (d *decoder) unmarshalMap(size uint, offset uint, result reflect.Value) (ui
...
-	return newOffset, err
+
+	switch result.Kind() {
+	default:
+		_, err := d.decode(pointer, result)
+		return newOffset, err
+	case reflect.Uintptr:
+		result.Set(reflect.ValueOf(uintptr(pointer)))
+		return newOffset, nil
+	}
```
Comment (oschwald) [inline decoder.go]: Is this change necessary? Wouldn't the same thing happen in the call to `.decode()` anyway?

**Comment by oschwald** [comment]
Overall this seems reasonable. I am a bit concerned with people not understanding the functionality and filing bug reports due to misuse. Longer term, I would like to implement some sort of caching strategy, although that might make more sense in a higher-level library like `geoip2` rather than in this library.

File: decoder.go (hunk 1 above)
Comment (jgraettinger) [inline decoder.go]: This is de-referencing the pointer at `offset` and storing the result, whereas the case in `decode` is storing the referenced pointer _prior_ to decode.

Comment (jgraettinger) [inline decoder.go]: I'll add comments here & above to clarify the difference.

Comment (oschwald) [inline decoder.go]: Right, but then it calls `decode` on the new offset from the pointer, which has the `Uintptr` handling for the new offset.

Comment (jgraettinger) [inline decoder.go]: Ah, quite right. Updated to remove.

**Comment by jgraettinger** [comment]
>  I am a bit concerned with people not understanding the functionality and filing bug reports due to misuse.

I've refactored and extended documentation on LookupOffset & Decode to document the behavior & the motivating use case. Let me know your thoughts.

>  It also seems to add a small amount of overhead in the benchmarks, but that might be unavoidable.

Here's what I'm seeing now:
```
master:
BenchmarkMaxMindDB-4       50000             33877 ns/op
BenchmarkCountryCode-4    500000              2225 ns/op

offset-lookups:
BenchmarkMaxMindDB-4       50000             35602 ns/op
BenchmarkCountryCode-4    500000              2224 ns/op
```

> Longer term, I would like to implement some sort of caching strategy, although that might make more sense in a higher-level library like geoip2 rather than in this library.

Agreed, I don't think that's appropriate in `maxminddb-golang` itself. Caching requires too much knowledge regarding the record type being extracted, and IMO it's a feature that `maxminddb` _doesn't_ impose such knowledge.

**Comment by oschwald** [comment]
Looks great. Thanks! :+1:
